core/mm/file.c

```diff
@@ -0,0 +1,168 @@
+// SPDX-License-Identifier: BSD-2-Clause
+/*
+ * Copyright (c) 2019, Linaro Limited
+ */
+
+#include <kernel/panic.h>
+#include <kernel/refcount.h>
+#include <mm/file.h>
+#include <mm/fobj.h>
+#include <stdlib.h>
+#include <string.h>
+#include <sys/queue.h>
+#include <types_ext.h>
+#include <util.h>
+
+/*
+ * struct file - file resources
+ * @tag:	Tag or hash uniquely identifying a the file
+ * @refc:	Reference counter
+ * @link:	Linked list element
+ * @num_slices:	Number of elements in the @slices array below
+ * @slices:	Array of file slices holding the fobjs of this file
+ *
+ * A file is constructed of slices which may be shared in different
+ * mappings/contexts. There may be holes in the file for ranges of the file
+ * that can't be shared.
+ */
+struct file {
+	uint8_t tag[FILE_TAG_SIZE];
+	unsigned int taglen;
+	struct refcount refc;
+	TAILQ_ENTRY(file) link;
+	unsigned int num_slices;
+	struct file_slice slices[];
+};
+
+static struct mutex file_mu = MUTEX_INITIALIZER;
+static TAILQ_HEAD(, file) file_head = TAILQ_HEAD_INITIALIZER(file_head);
+
+static int file_tag_cmp(const struct file *f, const uint8_t *tag,
+			unsigned int taglen)
+{
+	if (f->taglen != taglen)
+		return -1;
+	return memcmp(tag, f->tag, taglen);
+}
+
+static struct file *file_find_tag_unlocked(const uint8_t *tag,
+					   unsigned int taglen)
+{
+	struct file *f = NULL;
+
+	TAILQ_FOREACH(f, &file_head, link)
+		if (!file_tag_cmp(f, tag, taglen))
+			return f;
+
+	return NULL;
+}
+
+static void file_free(struct file *f)
+{
+	size_t n = 0;
+
+	for (n = 0; n < f->num_slices; n++)
+		fobj_put(f->slices[n].fobj);
+
+	free(f);
+}
+
+struct file *file_new(uint8_t *tag, unsigned int taglen,
+		      struct file_slice *slices, unsigned int num_slices)
+{
+	unsigned int s = 0;
+	unsigned int n = 0;
+	struct file *f = NULL;
+	bool did_insert = false;
+
+	if (taglen > sizeof(f->tag))
+		return NULL;
+	if (MUL_OVERFLOW(num_slices, sizeof(*slices), &s))
+		return NULL;
+	if (ADD_OVERFLOW(s, sizeof(*f), &s))
+		return NULL;
+	f = calloc(1, s);
+	if (!f)
+		return NULL;
+
+	memcpy(f->tag, tag, taglen);
+	f->taglen = taglen;
+	refcount_set(&f->refc, 1);
+	for (n = 0; n < num_slices; n++) {
+		f->slices[n].fobj = fobj_get(slices[n].fobj);
+		if (!f->slices[n].fobj ||
+		    ADD_OVERFLOW(slices[n].page_offset,
+				 slices[n].fobj->num_pages, &s))
```

Comment:
For sanity, maybe also check there is not already a matching slice, something like `assert(!file_find_slice(f, slices[n].page_offset))` before the slice is stored.

Reply:
This became a bit complicated and also highlight an error, I'll take this as a `[review]` patch on top.